pyexchange/fix.py

```diff
@@ -0,0 +1,292 @@
+#
+# Copyright (C) 2019 EdNoepel
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import asyncio
+import logging
+import simplefix
+import time
+import queue
+import threading
+
+from datetime import datetime, timedelta
+from enum import Enum
+from typing import List
+from concurrent.futures import ProcessPoolExecutor
+
+
+class FixConnectionState(Enum):
+    UNKNOWN = 0
+    DISCONNECTED = 1
+    CONNECTED = 2
+    LOGGED_IN = 3
+    LOGGED_OUT = 4
+
+
+def fprint(encoded_msg):
+    return encoded_msg.replace(b"\x01", b"|")
+
+
+class FixEngine:
+    """Enables interfacing with exchanges using the FIX (Financial Information eXchange) protocol.
+    This class shall implement common logic for connection management and fulfill relevant functions from PyexAPI.
+
+    Note that simplefix automatically populates fields 9 (message length) and 10 (checksum)."""
+
+    logger = logging.getLogger()
+    read_timeout = 30
+    write_timeout = 10
+    read_buffer = 128
+
+    def __init__(self, endpoint: str, sender_comp_id: str, target_comp_id: str, username: str, password: str,
+                 fix_version="FIX.4.4", heartbeat_interval=3):
+        assert isinstance(endpoint, str)
+        assert isinstance(sender_comp_id, str)
+        assert isinstance(target_comp_id, str)
+        assert isinstance(username, str)
+        assert isinstance(password, str)
+        self.endpoint = endpoint
+        self.senderCompId = sender_comp_id
+        self.targetCompId = target_comp_id
+        self.username = username
+        self.password = password
+        self.fix_version = fix_version
+        self.heartbeat_interval = heartbeat_interval
+        self.sequenceNum = 0
+        self.connection_state = FixConnectionState.DISCONNECTED
+
+        self.reader = None
+        self.writer = None
+        self.parser = simplefix.FixParser()
+
+        # This lock probably isn't needed because `reader.read` blocks.
+        self.lock = asyncio.Lock()
+        self.caller_loop = asyncio.get_event_loop()
+        self.session_loop = None
+        self.last_msg_sent = None
+        self.logging_out = False
+
+        self.write_queue = queue.Queue()
+        self.application_messages = queue.Queue()
+
+    async def _read_message(self):
+        """Reads the next message from the server"""
+        # await self.lock.acquire()
```

Comment:
I need to spend some time with the `asyncio` docs confirming these locks aren't really needed.

Reply:
`asyncio.StreamReader` and `StreamWriter` are not threadsafe.  Reads always initiate from the session thread, so no problem there.  Writes _usually_ initiate from a (threadsafe) standard library [Queue](https://docs.python.org/3.6/library/queue.html?highlight=queue#module-queue).  The only exception is `logout` which blocks on the caller's event loop while `_session_proc` keeps churning the `write_queue`.  As such, I removed the (commented-out) one from `_read_message` and uncommented the one in `_write_message`.

Reply:
Makes sense, thanks for the fix